### ISSUE TITLE
maxDataPoints handling

### ISSUE THREAD

**Issue by DanCech** [raintank/metrictank#463]
In mainline graphite, the finders return the highest-resolution aggregation level that covers the interval the user is querying, regardless of the maxDataPoints specified by the user. The maxDataPoints parameter is used on the final result of the query immediately before it's returned to the user, to resample the result if required to limit the total number of points.

In contrast, MT attempts to use the maxDataPoints parameter to choose the lowest-resolution aggregation level that contains at least maxDataPoints during the specified interval.  The assumption is that this is done to reduce the amount of data MT needs to return.

The problem with this approach is that it makes the results of many different kinds of queries dependent on the maxDataPoints parameter, especially if there is some intermediate processing being done before returning the final result set.  It's especially insidious when used with Grafana, because the maxDataPoints Grafana sends is dependent on the width of the graph, and thus the size of the browser window.

Side Note: this is true even for singlestat panels, which is a little strange.

In my opinion this should be changed to use the same logic as graphite when selecting the best aggregation level to serve a request, by selecting the highest-resolution interval that covers the requested time range.  This will ensure that any intermediate processing functions in graphite-api have access to the complete data, and it is only resampled down to the requested resolution (determined via maxDataPoints) after any processing has been performed.

**Comment by Dieterbe** [comment]
First some small corrections (they do not change your overall point)

> graphite ... to resample the result if required to limit the total number of points.

FWIW it consolidates (calls a summarisation function like avg/sum/min/max), it doesn't sample.

> MT attempts to use the maxDataPoints parameter to choose the lowest-resolution aggregation level that contains at least maxDataPoints during the specified interval. 

No. It will select the the highest-resolution aggregation level that contains no more than maxDataPoints points. in some cases it will still prefer to read from a higher resolution archive and apply consolidation on it.  For more details see https://github.com/raintank/metrictank/blob/master/docs/consolidation.md#the-request-alignment-algorithm

Now, to your main point.

> The problem with this approach is that it makes the results of many different kinds of queries dependent on the maxDataPoints parameter

Can you elaborate on why this matters or why this is a problem? I suspect the problem is that people have improperly set consolidateBy functions, but this would also be a problem with graphite.
Can you give an example of a graphite query and what kind of issue you get with it?

**Comment by DanCech** [comment]
> No. It will select the the highest-resolution aggregation level that contains no more than maxDataPoints points. in some cases it will still prefer to read from a higher resolution archive and apply consolidation on it.

That's even worse, since maxDataPoints usually represents the number of pixels on the final graph we want to err on the side of getting higher-resolution raw data and then resampling/consolidating/interpolating (pick your favorite term) to get the number of points the user requested.  At the MT level of the stack even doing that is most likely too much processing, and will cause issues if any aggregation functions are used in graphite-api.

> Can you elaborate on why this matters or why this is a problem? I suspect the problem is that people have improperly set consolidateBy functions, but this would also be a problem with graphite.
Can you give an example of a graphite query and what kind of issue you get with it?

`summarize()` is an interesting example, if we have a query like `summarize(scale(summarize(wp-billing.usage.1.checks_per_minute, '1h', 'avg', false), 60), '10y', 'sum', false)` executed over the last 30 days, then the results of summing the output values will look like:
```
maxDataPoints  Graphite  MT
1              214m      297k
10             214m      2.97m
100            214m      26.7m
1000           214m      106m
10000          214m      214m
100000         214m      214m
```

This is happening because the raw data is getting pre-consolidated by metrictank before it's returned to graphite, so when maxDataPoints is less than the number of hours in the interval the data is incorrect.

The result for 1000 is particularly interesting, since the consolidation causes in the result set for `scale(summarize(wp-billing.usage.1.checks_per_minute, '1h', 'avg', false), 60)` to have a null every other value because of the phase interaction of the consolidation to 1000 points by MT followed by the consolidation into 720 points by `summarize('1h', 'avg', false)`.

**Comment by woodsaj** [comment]
It is clear that by handling consolidation in MT, the results from a query to a stock graphite-web and our MT can be wildly different.

In most circumstances, the 'correct' summarize results can be obtained from MT by setting the correct consolidateBy.  So if you are doing `summarize(#A, '1h', 'sum',false)` then you should set consolidateBy('sum') so that if consolidation is needed you end up summing summed data.

The biggest concern that i have is that the  `summarize(scale(summarize(wp-billing.usage.1.checks_per_minute, '1h', 'avg', false), 60), '10y', 'sum', false)`  query can not be executed in MT at all due to the 2 summarize calls using different consolidation methods.  If you set consolidateBy(sum), then the 1h summaries will be the averages of sums instead of averages of averages.  

Additionally in MT, if the maxDataPoints and time range result in the data being consolidated to a resolution less than 1h, then the result will be wrong as the `scale(6)` expects the data to be at 1h resolution. Though in fairness the same problem would happen in Graphite-web if a rollup with a resolution of less than 1h was selected (it just so happens that in our deployment 1h data is the lowest resolution we have).


Though there is a discrepancy in the results here, i am not convinced that we should fix this prior to just moving graphite functions directly into MT.  Though this query does not work, the same/similar result can be obtained

* `summarize(1h, avg, false)`: is used to:
    a) handle all data resolutions selected from whisper (assumes that 1h is the lowest resolution data)
    b) treat missing data (nulls) as the hourly average rather then 0.  `scale(avg(null,1,1,1) , 4)` = 4 whereas  `sum(null,1,1,1)`=3
* scale(60): is needed to convert 1h average check_per_minute rate to a total number of checks in the 1h summary.
* summarize(10y, sum, false): is used to reduce the data to a single value which is the total number of checks performed in the requested time range.

The tricky component of this query is the handling of null data.  If that is a hard requirement then we need to just use

`scale(wp-billing.usage.1.checks_per_minute, <number of minutes in requested time range>)&maxDataPoints=1`

If we are ok with treating null as zero, then the following will work with all time ranges.
`consolidateBy(wp-billing.usage.1.checks_per_minute, 'sum')&maxDataPoints=1`

**Comment by woodsaj** [comment]
after further thought, i do think we should fix this. as compatibiliy/correctness with stock graphite is important. 
Though a lot of work has been put into consolidation logic in MT, i do feel that sending the highest resolution data available to the graphite functions is the correct thing to do.   

This will have a noticeable performance penalty for our deployments due to graphite-api needing to process more data, but we can handle that with more hardware until graphite functions are moved into MT.

**Comment by DanCech** [comment]
Glad to hear it, from a user's perspective I can definitely say that while it's possible to work around the issues it's definitely an exercise in frustration trying to figure out why you're getting different results based on the maxDataPoints setting (which is usually dictated by the width of the graph), and why in some situations graphite will work just fine for singlestat graphs with maxDataPoints=1 whereas MT won't.

> If we are ok with treating null as zero, then the following will work with all time ranges.
> `consolidateBy(wp-billing.usage.1.checks_per_minute, 'sum')&maxDataPoints=1`

That will work assuming that the collector is reporting at 1-minute intervals, if the collector reports the per-minute rate every 30 seconds then the total would be double the correct value.  The special handling of consolidateBy here is also a departure from how graphite handles things, but I can understand the reasoning behind it, and I think it's likely to give users results that are closer to what they expect.

**Comment by Dieterbe** [comment]
we just had a meeting about this with @DanCech and @replay 
seems to me MT returning the highest-res to graphite is the right way to go.
the problem manifests itself in two main ways
- MT overeagerly consolidating so that input to graphite is too coarse
- misalignment between consolidated points from MT and what graphite wants (e.g. graphite wants to do summarize in hour-buckets, but Mt may have consolidated points so that they are also hourly, but not on the hour buckets)

Dan also proposed an idea of how to make it easier for users to communicate to the tsdb which aggregation they want (since consolidateBy is too obscure). which is using `-avg`, `-min` etc suffixes to series names. the graphite finder/MT could even present those as if they exist as real metrics, so that things just work in grafana's query editor.  I like this idea since it makes people more aware of the importance of this topic, and incentives them to select a proper value.  But seems like this idea deserves to have its own design discussion irrespective of this ticket.

**Comment by replay** [comment]
@Dieterbe @DanCech and me just had a long meeting to discuss this. To summarize:

The problem is rooted in the fact that metrictank is smart by applying consolidations to data before returning it to graphite-api, which has the advantage that it reduces the amount of data to be sent over the wire, but it introduces problems because there are many different ways how the data can be consolidated and metrictank doesn't necessarily know what graphite-api needs for a given request.

As a solution we agreed on that metrictank should always send the rawdata of the aggregation level which provides the maximum granularity and still has all the data for the requested time range. 

This will lead to more data being returned to graphite-api, so before implementing and deploying it we want to know how much data that will be. So as a first step we should add a new metric to metrictank which for each request gives us the number of data points that would be returned to graphite-api if metrictank would have selected the aggregation level that provides maximum granularity (without changing anything else yet).

Once we see how many datapoints would be transferred to graphite-api, we can still decide if we really want to change metrictank to always return the maximum granularity. 

Furthermore we discussed that the metrictank finder should return every leaf node multiple times, once without suffix, and once for each consolidation method suffixed with a string like `-max`, `-min` etc. That way a user can select the consolidation method that should be used by metrictank if it has to return consolidated data if the raw data doesn't cover all of the requested time range.

**Comment by Dieterbe** [comment]
One other reason why the code currently is the way it is, is to assure, in basic cases (of just querying one or more series, or avgSeries(..), sumSeries(...), groupByNode(series), etc that we read rolled up data made with the same consolidation function as what is applied at runtime. (see https://blog.raintank.io/25-graphite-grafana-and-statsd-gotchas/#runtime.consolidation) with standard graphite you could consolidate at runtime with a different function then what the archive says, and get wrong results. with current MT, and in this scenario, you get correct data.
But as we're seeing when applying functions like perSecond, summarize, etc this becomes wrong.

So in addition to feeding graphite's processing functions with the highest res data we have, there's still the problem of which rollup archive to read from when we have to read from a rolled up archive. In graphite's case it's easy, there's only one, and when the user building charts is familiar with the carbon configuration and knows what he's doing (and uses a proper consolidateBy, matching what the server config says), then they can get a correct result from graphite more easily. To increase our compatibility with graphite, perhaps we should introduce a consolidation function in the rollup config as well, we'd use it as the one to pick by default, until we introduce the `-min`, `-max` syntax DanC suggested.
potentially we could deduce the default consolidateBy from it as well
we could also go via metrics2.0 and instead of having people make `pattern -> rollup function` rules, they could make `pattern -> mtype` rules (mtype is gauge vs count vs counter, etc) and we can set the correct default rollup function from that.

**Comment by Dieterbe** [comment]
@woodsaj @DanCech @replay  feedback on the above, in particular the step plan? thanks.

**Comment by DanCech** [comment]
I think the plan makes sense, and the more I think about it the more sense the `-min` approach makes as a way for the user to signal which rollup method they want.  Trying to implement it as a function would require some extra mechanism to pass it through to the backend vs just requesting the "synthetic" series.

**Comment by replay** [comment]
That makes sense.
The only thing I'm not sure about is how you imagine step 3 to look like. Do you mean you'd want to add additional config files to Metrictank that look like graphite's storage-schemas.conf and storage-aggregation.conf? Isn't that confusing because at the moment we only support one retention configuration and are not able to deal with graphite's way of configuring the retention policy per pattern.

**Comment by woodsaj** [comment]
I think you are going to need to change the order of your steps.
step 2, depends on step3.

But otherwise this looks like a good approach.

**Comment by Dieterbe** [comment]
I've been wanting to redo our archive-related config settings for a while, since there is no clear link between the global ones ( chunkspan, numchunks, ttl) and the one in agg-settings (which is a string of the form `aggSpan:chunkSpan:numChunks:TTL[:ready as bool. default true]`)
the way i had envisioned it is that we would configure the global stuff along with the rollup stuff, as they are all archives (esp in the code we talk a lot about "archives" which could be raw or rolled up. similar to whisper)

so it would be something along the lines of

```
# metrictank.ini
[archive-raw]
chunkspan = 10min
numchunks = 7
ttl = 35d

[archive-rollup]
interval = 5min
chunkspan = 1h
numchunks = 2
ttl = 3mon

[archive-rollup]
interval = 1h
chunkspan = 6h
numchunks = 2
ttl = 1y
ready = false
```
(note no interval setting for raw. we could add one which could serve as default in case interval is not passed with the data)

But now because:
1) rollups, and ostensibly also raw settings, will be per metric pattern, so there will be potentially quite a few more of these definitions, which means it would be nicer to write them down more compactly (using less lines)
2) we wish to make things more easy for people coming from graphite.


I'm thinking perhaps we should use, and extend the graphite format, which for the record uses 2 files like so:

```
# storage-aggregation.conf
[all_min]
pattern = \.min$
xFilesFactor = 0.1
aggregationMethod = min

# storage-schemas.conf
[default]
pattern = .*
retentions = 1s:1d

[apache_busyWorkers]
pattern = ^servers\.www.*\.workers\.busyWorkers$
retentions = 15s:7d,1m:21d,15m:5y
```
basically I think we can just extend the storage-schemas.conf `retention` field with a field for chunkspan and numchunks, and ready for all rollup archives.

that would mean we remove those settings from the metrictank config, and use 3 config files instead.

**Comment by woodsaj** [comment]
I agree that we should use the carbon storage-aggregation.conf and storage-schemas.conf with additional fields added for chunkspan and numchunks

cant we just interpret `aggreagtionMethod` to mean the default consolidation function to use for queries.

**Comment by replay** [comment]
I like the idea of defining aggregation methods as you showed `aggregationMethod = min,max,avg`. The advantage of this is that it fits well together with the whisper import procedure because in whisper files we only have one type of aggregation per archive, so if we'd import a whisper file where the aggregation method `X` is used we can make metrictank make sense of that by configuring `aggregationMethod = X`. 
The last time when we discussed that problem that in whisper files we don't have all the possible aggregations we were thinking about if it would make sense to just copy the data the whisper file contains into all possible aggregations or leave the ones we don't have empty. By being able to only enable certain aggregations in Metrictank this question would be answered.

**Comment by Dieterbe** [comment]
> cant we just interpret aggreagtionMethod to mean the default consolidation function to use for queries.

we could. it's just a bit misleading. we store many more aggregations that we want to encourage people to be aware of and use. However:
- by sticking to the one function notation that gets hidden
- by keeping the config tied to graphite, people will subconsciously make connections in their head and assume graphite characteristics apply for MT as well.

Perhaps we should make it so that only the listed functions are what is stored, the first one is what is used for queries by default. (slightly different from my previous proposal because then it was assumed if only 1 is mentioned, we still store all of them). `[*]`

The counterargument to this proposal is: if people still need to make adjustments to their config files, (add extra aggregationMethods, add numChunks and chunkSpan fields), then we may as well have them "convert" into whatever config file/syntax we want (including putting everything in metrictank.conf).

I'm not sold on it yet, but my gut says we should do this:
* make metrictank a nice experience with *unmodified* carbon config files. i.e. allow not specifying numchunks and chunkspans (we can come up with default values based on interval). Implement `[*]` above, meaning we only store the aggregations that are specified (so with unmodified files, that means only 1)
* encourage people (via quickstart guide etc) to put some thought into their aggregationMethods and specify multiple
* make this concept a first class citizen down the road: MT is aware of which aggregations can be queried, and this can be exposed to Grafana and made much nicer in the query editor. For any given series, it could be 1, a few, or all of them. But it always depends on which metric(s) you're looking at.

**Comment by replay** [comment]
You're right, we'll have to handle the case when `storage-aggregations.conf` changes over time. What i had in mind was the most simple case, if your point `[*]` gets implemented and only the listed aggregations get stored, then somebody simply imports whisper files and copies the according `storage-aggregations.conf` (which only list one aggregation in the old format) the data fits what's configured in the config file.
So further down the road if we for example implement the `-min` and `-max` syntax described in https://github.com/raintank/metrictank/issues/463 we could only return the configured aggregations in `/find` calls so the auto complete doesn't even show empty unconfigured aggregations. 
This leads to much less confusion about why some aggregations are empty and some are not, the auto complete in grafana is basically showing that only some are there.

**Comment by Dieterbe** [comment]
right.
FTR during the hosted call we decided to implement my last proposal above.

**Comment by Dieterbe** [comment]
this is resolved now via the PR's linked above.
feel free to re-open if I missed something.
